### PR TITLE
[Train] Preemption-aware worker lifecycle for Ray Train v2

### LINKED ISSUE #63968: [RFC][Train] Preemption-aware lifecycle for Ray Train
# [RFC][Train] Preemption-aware worker lifecycle for Ray Train

## Problem Statement

Node preemption is common in modern training. Spot/preemptible capacity is 60–90% cheaper than on-demand, but in a fixed worker shape with standard checkpoint-and-restart, **every preemption wastes `N_workers × (time since last checkpoint)` accelerator-hours.**

Cloud providers give advance warning before reclaiming a node — [AWS spot up to 120s](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-instance-termination-notices.html), [GKE ~30s](https://docs.cloud.google.com/kubernetes-engine/docs/concepts/spot-vms#termination-graceful-shutdown). Ray Train ignores that warning and reacts only to the *actual* death of the worker, after it surfaces as a `RayActorError`. This produces the **"checkpoint every 1000 steps, died at step 999"** problem: a full checkpoint interval of progress is thrown away on every preemption, even though the system had tens of seconds of advance notice.

Walking the preemption lifecycle, there are three distinct gaps:

| Phase | Gap |
|-------|-----|
| **Before preemption** | The advance-warning window is wasted. The signal exists in Ray Core (`get_draining_nodes()` carries the node + deadline) but is never surfaced to user code, so there is no chance to checkpoint just-in-time, hand off in-flight work, or exit cleanly. |
| **At preemption** | Surviving workers are torn down together with the doomed ranks. They lose in-flight work, and they lose warm state (weights in accelerator memory, runtime-env installs) — forcing a full cold-start rebuild. |
| **After preemption** | The failure class is invisible to the retry policy. A spot reclaim, an OOM, and a code bug all collapse into the same `WorkerGroupError` and consume the same `max_failures` budget. |

With the three gaps above, to survive routine spot churn, a user must set `max_failures` high — but that also means a real bug retries many times instead of failing fast. Preemption and real failures need independent budgets.

**Goal:** 
1. make each of the three phases addressable from user code and/or framework policy, 
2. Backward compatible. A UDF that never touches the new API must behave exactly as it does today.

## Current State

Ray Train already has a mature recovery toolkit. The key point is that **every mechanism is reactive — it activates only *after* a worker has already died, and treats all deaths identically.**

- **Checkpoint + `FailureConfig(max_failures=N)`** — the baseline. On a `RayActorError`, the controller tears down the worker group and rebuilds at the same world size; the run restores from the last checkpoint reported via `ray.train.report()`.

Image

- **Elastic training, `num_workers=(min, max)`** — lets training rebuild at a smaller world size if survivors `≥ min_workers`, while preempted nodes are re-provisioned.
- **TorchFT integration (in progress)** — per-step quorum: survivors keep stepping with a smaller replica set and replacements recover from peer memory, eliminating the disk-checkpoint dependency for the in-quorum case.

None of these uses the advance warning, lets survivors finish in-flight work before teardown, or distinguishes a planned drain from a real crash. This RFC adds the *before* and the *after* on top of that foundation rather than replacing any of it.

## Proposed Approach

Three additive components, one per gap:

| Gap | Component |
|-----|-----------|
| Before — warning window wasted | **`PreemptionWatcher`** surfaces the cloud's drain signal into the UDF via `ray.train.preemption_status()`, so users can JIT-checkpoint, hand off work, clean up, or exit cleanly. |
| At — survivors torn down with the doomed ranks | **`PreemptingState`** — a new controller state that gives survivors a configurable drain window to finish their step and flush an in-flight checkpoint before the group is rebuilt. |
| After — failure class invisible | **`PreemptionError`** — a distinct failure class with its own retry budget (`max_preemption_failures`), independent of `max_failures`. |

Image

### User-facing API

A single function, queried at any step boundary, returns `None` in steady state:

```python
@dataclass(frozen=True)
class PreemptionInfo:
    deadline_ms: float                   # UNIX timestamp; pod expected to be reclaimed by this time
    preempted_ranks: list[int]      # all worker ranks whose hosts are draining
    preempted_node_ids: list[str]   # corresponding node IDs

def get_preemption_info() -> Optional[PreemptionInfo]:
    ...
```

A representative UDF — preempted ranks clean up; survivors capture a just-in-time checkpoint:

```python
def train_loop_per_worker(config):
    ckpt = ray.train.get_checkpoint()
    if ckpt:
        ...  # restore model + data-iterator state from the latest checkpoint

    rank = ray.train.get_context().get_world_rank()
    for step, batch in enumerate(train_iterator):
        info = ray.train.get_preemption_info()
        if info is not None:
            if rank in info.preempted_ranks:
                cleanup_local()                          # skip pointless I/O on a doomed host
                ray.train.report({"step": step})
            else:
                # survivor: capture rank-0 data state + write a JIT checkpoint
                ds_state = ray.train.collective.broadcast_from_rank_zero(
                    train_iterator.state_dict()
                )
                save_jit_checkpoint(model, ds_state, step)
            return                                       # clean UDF exit
        train_step(model, batch)
```

### Signal flow and end-to-end timing

1. **Cloud marks one or more nodes as draining.** Ray Core surfaces this via `get_draining_nodes()` within a few seconds.
2. **`PreemptionWatcher` detects.** A small actor co-located with the controller polls `get_draining_nodes()` every `poll_interval_s`. On change, it maps drained node IDs to affected ranks via a failure-domain map (below), resolves the deadline, and builds a `PreemptionInfo`.
3. **Fan-out to workers.** The watcher calls `mark_preempt(info)` directly on every worker actor — both doomed and survivor — so survivors learn they're survivors. Workers store the info in their `TrainContext` (where `preemption_status()` reads it) and echo it in `WorkerStatus.preemption_info`.
4. **UDF observes** via `preemption_status()` at the next step boundary and branches.
5. **Controller transitions to `PreemptingState`** on its next health-check poll, when it sees any worker reporting `preemption_info != None`.

We prefer the watcher fanning out **directly to workers** (rather than watcher → controller → workers) so that workers act inside the warning window without waiting for a controller loop iteration. End-to-end budget, from a cloud signal landing to each consumer reacting:

| Step | Latency |
|------|---------|
| Cloud signal lands in `get_draining_nodes()` (Ray Core propagation) | ~0–2s |
| Watcher's poll picks it up | up to `poll_interval_s` (default 5s) |
| Watcher fan-out `mark_preempt` (direct RPC, parallel) | <1s |
| → worker UDF sees info at next step boundary | ~6s + remaining step time |
| Controller's next poll observes the echo and enters `PreemptingState` | up to 2s |

With a typical 30s drain notice, that leaves ~20s of headroom before SIGKILL — enough for the UDF to JIT-checkpoint, survivors to flush, and the controller to orchestrate the transition.

### Failure-domain mapping

Drain signals are per-node, but the unit that must be flagged atomically is the **failure domain**:

- **TPU slice:** TPU slices are fate-shared and atomically preempted, but the hosts within a slice are not drained simultaneously. Ray Core tags every TPU node with a slice label (`RAY_NODE_TPU_SLICE_NAME_KEY`); all ranks sharing that label form one domain. Draining *any* host in a slice flags *every* rank in that slice from the first signal.
- **Otherwise:** the failure domain is the node itself (standard GPU/CPU clusters).

The watcher builds `{node_id: [ranks_in_same_failure_domain]}` once per worker group so that we fan out correct information to all the udf threads. 

### PreemptionWatcher lifecycle (`PreemptionCallback`)

A `WorkerGroupCallback` owns the watcher: it spawns a fresh watcher in `after_worker_group_start` (building the failure-domain map and handing over worker actor handles) and stops it in `before_worker_group_shutdown`. One watcher per worker group means each restart gets a clean watcher and a fresh map — and it composes with elastic training for free.

Two deliberate choices:
- **Separate actor for isolation**, and so the drain *source* can later be pluggable — a developer API that accepts notices from a K8s informer or cloud event queue, not only `get_draining_nodes()`.
- **Not enabled with TorchFT initially.** TorchFT already handles peer loss via per-step quorum (survivors keep stepping), and recovers replacements from peer memory — which removes the need for JIT checkpointing. Driving a `PreemptingState` transition on top would interfere with its graceful-degrade path.

### PreemptingState and exit criteria

When the controller observes preemption, it enters `PreemptingState` instead of tearing down immediately. During this state it keeps the replacement resource request asserted so node provisioning overlaps with survivor cleanup. It leaves the state when **`_should_leave_preempting(info)`** is true:

1. every rank has terminated (returned cleanly, raised, or was SIGKILL'd), **or**
2. `info.seconds_remaining <= 0` (deadline expired — safety net).

On exit it synthesizes a `PreemptionError` and hands it to the failure policy.

### PreemptionError and failure classification

```python
class PreemptionError(TrainingFailedError):
    """Worker group torn down due to node preemption.

    Distinct from WorkerGroupError so planned preemption consumes a separate
    retry budget (FailureConfig.max_preemption_failures, default -1 = unlimited)
    instead of max_failures, which is reserved for real failures (OOM, hardware
    fault, user-code bugs).
    """
    preempted_node_ids: list[str]
    preempted_ranks: list[int]
    deadline_exceeded: bool              # True if we left via the deadline path
    worker_failures: dict[int, Exception]  # rank -> underlying error, for diagnostics
```

Classification at `PreemptingState` exit:

```python
if any(is_preemption_caused(e) for e in errors):   # ≥1 preemption-caused error / cascade
    -> PreemptionError      # consumes max_preemption_failures (default ∞)
elif errors:                                        # errors, none preemption-caused
    -> WorkerGroupError     # consumes max_failures (default 0 -> fail fast)
else:                                               # clean / forced teardown
    -> PreemptionError      # synthesized
```

`FailureConfig` gains `max_preemption_failures` (default `-1` = unlimited), consumed independently of `max_failures` (default `0`):

```python
FailureConfig(max_failures=3, max_preemption_failures=-1)   # fail fast on bugs, ride out spot churn
```

`DefaultFailurePolicy.make_decision` adds a `PreemptionError` case. A user on the default config gets retry-on-preemption automatically with no special configuration, while real bugs still fail fast — which is what most users actually want.

### The `report()` barrier — open question

`ray.train.report()` is a per-step collective: the report handler waits until **every** worker has produced a `TrainingReport` for the step before consolidating and persisting the checkpoint. Under preemption this is a problem: a doomed rank may be SIGKILL'd before it reaches `report()` (a 5–10s drain notice plus a long in-flight step can leave it stuck in `train_step()`). If survivors are waiting at the barrier for a report that never comes, a perfectly valid survivor checkpoint is discarded when the group tears down.

Three positions, all viable:

| Option | Mechanism | Trade-off |
|--------|-----------|-----------|
| **1. Strict barrier + `report(None)`** | Doomed ranks must call `report(metrics_only)` before exiting so the handler still sees N reports. | Zero framework change, but fragile — depends on the doomed UDF reaching `report()` in time, and fails silently if it doesn't. |
| **2. Relax the barrier in `PreemptingState`** | Report handler expects reports only from survivor ranks; doomed ranks are ignored. | Doomed ranks can SIGKILL freely; preempted-branch user code becomes optional. Requires plumbing controller state into the handler, and `expected_ranks` must be **snapshotted at state entry** (not recomputed) so a staggered drain doesn't orphan an in-flight survivor checkpoint. |
| **3. Synthesize empty reports** | Controller injects a synthetic empty `TrainingReport` for any preempted rank that exited without reporting. | Barrier semantics unchanged, handler stays stateless. Controller must track per-rank report state and guard against late/duplicate reports; synthetic reports are a fiction in debug logs. |

The right answer interacts with checkpoint topology. We analyzed save-time × topology (pure DDP, model-parallel with a global process group, model-parallel with per-group PGs, and raw per-rank `torch.save`):

- **Pure DDP** (the dominant case): relaxing the barrier (Option 2) is a clean win — survivors hold a complete model replica, so their JIT checkpoint is fully recoverable.
- **Model-parallel with a library checkpoint** (Megatron / DeepSpeed / TorchTitan): the library's own commit marker / global collective dominates the `report()` barrier and safely catches partial writes — marginal difference between options.
- **Raw per-rank `torch.save` with model-parallel sharding**: relaxing the barrier can commit a directory missing the doomed ranks' shards, recoverable only with a topology-aware loader. Here Option 1's stricter behavior is safer.

## Backward Compatibility

A UDF that never calls `preemption_status()` is unaffected by the API, but still benefits from the classification:

| Aspect | Today | With this design (no API use) |
|--------|-------|-------------------------------|
| Failure classification | preemption ≡ OOM ≡ crash | preemption tagged as `PreemptionError` |
| Retry budget | `max_failures` (often 0 → job fails) | `max_preemption_failures` (default ∞ → retries) |
| User code changes | — | **none** |
| Lost work on retry | one checkpoint interval | one checkpoint interval (same) |

The watcher still populates each worker's context (inert, since nothing reads it), the controller still enters `PreemptingState`, and on eviction the run retries from the last periodic checkpoint. No correctness violation, no manual recovery.

## Out of Scope / Future Work

- **Suspend-and-resume** — avoid tearing down survivors at all; survivors retain in-memory state and only the replacement pays cold start, resuming from the last *completed* step rather than the last *checkpointed* step. For TorchTrainer this is largely covered by the TorchFT integration (in-quorum case); for JaxTrainer it is blocked on resizable Placement Groups in Ray Core and JAX-on-TPU runtime support. JIT checkpointing (this RFC) already captures the dominant cost — lost progress — so suspend-and-resume is a follow-up, not a prerequisite.

## References

- Prototype: [PR #63475](https://github.com/ray-project/ray/pull/63475)
- [AWS spot instance termination notices](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-instance-termination-notices.html) · [GKE spot VM graceful shutdown](https://docs.cloud.google.com/kubernetes-engine/docs/concepts/spot-vms#termination-graceful-shutdown)
- [JAX fault-tolerance APIs](https://docs.jax.dev/en/latest/fault_tolerance.html#id4) (relevant to JaxTrainer suspend-and-resume)

## Environment

- Ray version: master
- Component: Ray Train v2 (`python/ray/train/v2/`)

### PR DESCRIPTION
**Pull request by Truc54**
## Description
This PR implements a preemption-aware worker lifecycle in Ray Train v2 to handle node preemption events gracefully. Instead of ignoring node preemption signals and only reacting to unplanned actor deaths (after it surfaces as a `RayActorError` and losing all progress since the last checkpoint), Ray Train will now actively poll node draining status from Ray Core's `get_draining_nodes` GCS RPC. 

Upon receiving a preemption signal, the controller transitions to a transient `PreemptingState` and grants the surviving nodes a grace window to finish their current training loop step, write a final just-in-time (JIT) checkpoint, and exit cleanly. Preemptions are tracked under an independent retry budget controlled by `max_preemption_failures` in `FailureConfig`.

## Related issues
Closes #63968.

### Key Changes
- **Public APIs**:
  - `ray.train.preemption_status()`: Public helper function for workers to check if a preemption event has been signaled.
  - `PreemptionInfo` dataclass: Holds the preemption deadline, reason, ranks preempted, and node IDs.
  - `PreemptionError` exception: Independent failure class representing worker group preemption.
  - `max_preemption_failures` in `FailureConfig`: Separate retry budget for preemption failures (independent from general unplanned failures).
- **Core Signal Flow & Routing**:
  - `PreemptionCallback` / `PreemptionWatcher`: Polling mechanism querying Ray Core node draining status and mapping affected nodes to their world ranks.
  - TPU slice failure domain detection to preempt the whole slice if one node is drained.
  - Fan-out preemption signal via `mark_preempt()` RPC to affected worker actors.
- **Controller State Machine**:
  - Transitions to `PreemptingState` upon receiving a preemption signal.
  - Implements grace window logic: waits until all preempted nodes exit cleanly or the deadline expires.
  - Synthesizes `PreemptionError` for clean-exit preempted nodes to trigger appropriate failure handling.

### Verification
- **Unit Tests**:
  - Created a comprehensive test suite in `python/ray/train/v2/tests/test_preemption_unit.py` with custom import hook mocking to verify state machine transitions, failure policies, deadline resolver logic, and TPU slice failure domain configurations without a real compiled Ray cluster. All 24 unit tests pass successfully.
- **Integration Demo**:
  - Implemented `python/ray/train/v2/tests/preemption_integration_demo.py` showcasing multi-node local cluster behavior, node draining injection via GCS client, JIT-checkpointing by survivors, and subsequent state restoration.

### CODE REVIEW

**Comment by liulehui** [comment]
Hi thanks for the contribution! This feature is actively being implemented by the Ray Train team, if this feature resonates, please check our next few releases!
